File: avalon-modules/avalon-dataset/avalon-dataset-controller/src/test/java/io/github/lishangbu/avalon/dataset/controller/MoveDamageClassControllerTest.java
package io.github.lishangbu.avalon.dataset.controller;

import static org.junit.jupiter.api.Assertions.assertSame;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import io.github.lishangbu.avalon.dataset.entity.MoveDamageClass;
import io.github.lishangbu.avalon.dataset.service.MoveDamageClassService;
import java.util.List;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;

@ExtendWith(MockitoExtension.class)
class MoveDamageClassControllerTest {
    @Mock private MoveDamageClassService moveDamageClassService;
    @InjectMocks private MoveDamageClassController moveDamageClassController;

    @Test
    void getMoveDamageClassPage_delegatesToService() {
        MoveDamageClass moveDamageClass = new MoveDamageClass();
        Pageable pageable = PageRequest.of(0, 5);
        Page<MoveDamageClass> page = new PageImpl<>(List.of(moveDamageClass));
        when(moveDamageClassService.getPageByCondition(any(MoveDamageClass.class), eq(pageable)))
                .thenReturn(page);

        Page<MoveDamageClass> result =
                moveDamageClassController.getMoveDamageClassPage(pageable, moveDamageClass);

        assertSame(page, result);
        verify(moveDamageClassService).getPageByCondition(moveDamageClass, pageable);
    }

    @Test
    void listMoveDamageClasses_delegatesToService() {
        MoveDamageClass moveDamageClass = new MoveDamageClass();
        List<MoveDamageClass> list = List.of(moveDamageClass);
        when(moveDamageClassService.listByCondition(moveDamageClass)).thenReturn(list);

        List<MoveDamageClass> result =
                moveDamageClassController.listMoveDamageClasses(moveDamageClass);

        assertSame(list, result);
        verify(moveDamageClassService).listByCondition(moveDamageClass);
    }

    @Test
    void save_delegatesToService() {
        MoveDamageClass moveDamageClass = new MoveDamageClass();
        when(moveDamageClassService.save(moveDamageClass)).thenReturn(moveDamageClass);

        MoveDamageClass result = moveDamageClassController.save(moveDamageClass);

        assertSame(moveDamageClass, result);
        verify(moveDamageClassService).save(moveDamageClass);
    }

    @Test
    void update_delegatesToService() {
        MoveDamageClass moveDamageClass = new MoveDamageClass();
        when(moveDamageClassService.update(moveDamageClass)).thenReturn(moveDamageClass);

        MoveDamageClass result = moveDamageClassController.update(moveDamageClass);

        assertSame(moveDamageClass, result);
        verify(moveDamageClassService).update(moveDamageClass);
    }

    @Test
    void deleteById_delegatesToService() {
        moveDamageClassController.deleteById(1L);
        verify(moveDamageClassService).removeById(1L);
    }
}

File: avalon-modules/avalon-dataset/avalon-dataset-controller/src/test/kotlin/io/github/lishangbu/avalon/dataset/controller/MoveDamageClassControllerTest.kt
package io.github.lishangbu.avalon.dataset.controller

import io.github.lishangbu.avalon.dataset.entity.dto.MoveDamageClassSpecification
import io.github.lishangbu.avalon.dataset.entity.dto.MoveDamageClassView
import io.github.lishangbu.avalon.dataset.entity.dto.SaveMoveDamageClassInput
import io.github.lishangbu.avalon.dataset.entity.dto.UpdateMoveDamageClassInput
import io.github.lishangbu.avalon.dataset.service.MoveDamageClassService
import org.babyfish.jimmer.Page
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertSame
import org.junit.jupiter.api.Test
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable

class MoveDamageClassControllerTest {
    @Test
    fun getMoveDamageClassPage_delegatesToService() {
        val service = FakeMoveDamageClassService()
        val controller = MoveDamageClassController(service)
        val pageable = PageRequest.of(0, 5)
        val page: Page<MoveDamageClassView> = Page(listOf(MoveDamageClassView("1", "physical", "物理", "desc")), 1, 1)
        service.pageResult = page
        val specification = MoveDamageClassSpecification(id = "1", internalName = "physical", name = "物理", description = "desc")

        val result = controller.getMoveDamageClassPage(pageable, specification)

        assertSame(page, result)
        assertSame(pageable, service.pageable)
        assertEquals("1", service.pageCondition!!.id)
    }

    @Test
    fun listMoveDamageClasses_delegatesToService() {
        val service = FakeMoveDamageClassService()
        val controller = MoveDamageClassController(service)
        val list = listOf(MoveDamageClassView("1", "physical", "物理", "desc"))
        service.listResult = list
        val specification = MoveDamageClassSpecification(id = "1", internalName = "physical", name = "物理", description = "desc")

        val result = controller.listMoveDamageClasses(specification)

        assertSame(list, result)
        assertEquals("1", service.listCondition!!.id)
    }

    @Test
    fun save_delegatesToService() {
        val service = FakeMoveDamageClassService()
        val controller = MoveDamageClassController(service)
        val command = SaveMoveDamageClassInput("physical", "物理", "desc")
        service.saveResult = MoveDamageClassView("1", "physical", "物理", "desc")

        val result = controller.save(command)

        assertSame(service.saveResult, result)
        assertSame(command, service.savedCommand)
    }

    @Test
    fun update_delegatesToService() {
        val service = FakeMoveDamageClassService()
        val controller = MoveDamageClassController(service)
        val command = UpdateMoveDamageClassInput("1", "physical", "物理", "desc")
        service.updateResult = MoveDamageClassView("1", "physical", "物理", "desc")

        val result = controller.update(command)

        assertSame(service.updateResult, result)
        assertSame(command, service.updatedCommand)
    }

    @Test
    fun deleteById_delegatesToService() {
        val service = FakeMoveDamageClassService()
        val controller = MoveDamageClassController(service)

        controller.deleteById(1L)

        assertEquals(1L, service.removedId)
    }

    private class FakeMoveDamageClassService : MoveDamageClassService {
        var pageCondition: MoveDamageClassSpecification? = null
        var pageable: Pageable? = null
        var listCondition: MoveDamageClassSpecification? = null
        var savedCommand: SaveMoveDamageClassInput? = null
        var updatedCommand: UpdateMoveDamageClassInput? = null
        var removedId: Long? = null

        var pageResult: Page<MoveDamageClassView> = Page(emptyList(), 0, 0)
        var listResult: List<MoveDamageClassView> = emptyList()
        lateinit var saveResult: MoveDamageClassView
        lateinit var updateResult: MoveDamageClassView

        override fun getPageByCondition(
            specification: MoveDamageClassSpecification,
            pageable: Pageable,
        ): Page<MoveDamageClassView> {
            pageCondition = specification
            this.pageable = pageable
            return pageResult
        }

        override fun save(command: SaveMoveDamageClassInput): MoveDamageClassView {
            savedCommand = command
            return saveResult
        }

        override fun update(command: UpdateMoveDamageClassInput): MoveDamageClassView {
            updatedCommand = command
            return updateResult
        }

        override fun removeById(id: Long) {
            removedId = id
        }

        override fun listByCondition(specification: MoveDamageClassSpecification): List<MoveDamageClassView> {
            listCondition = specification
            return listResult
        }
    }
}
